AX: use one AXObjectCache per frame

minorninth opened the pull request:
#### 167880ef7dd67803a1145e76f69a15df33d13219
<pre>
AX: use one AXObjectCache per frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=298901">https://bugs.webkit.org/show_bug.cgi?id=298901</a>
<a href="https://rdar.apple.com/160643943">rdar://160643943</a>

Reviewed by Tyler Wilcock and Joshua Hoffman.

Previously we had one AXObjectCache per page, spanning all frames
on that page. That design presented some challenges when trying to
implement support for remote frames - what is the root of the AXObjectCache
when the root frame is remote?

This patch refactors everything so that there&apos;s a separate AXObjectCache
for each LocalFrame. The AXObjectCache is owned by the document.

A class, AXLocalFrame, is added to parallel AXRemoteFrame.
It&apos;s a node in the accessibility tree, but its wrapper points
directly to the root of the corresponding child frame. This mimics
how AXRemoteFrame works.

Note that both with and without site isolation, the
AccessibilityScrollView class appears twice in each frame: it represents
the root of each frame, and there&apos;s also one instance of AccessibilityScrollView
for each child frame. To distinguish between these more easily, I added a new
role FrameHost for the scroll view in the parent frame.

As an example, suppose we have a simple document with an iframe in it. The
AX tree looks like this, with a total of three AccessibilityScrollView
instances.

- Parent frame

AccessibilityScrollView (Role::ScrollArea)
    AccessibilityRenderObject (Role::WebArea)
        ....
        AccessibilityRenderObject (corresponding to IFrame element, ignored)
            AccessibilityScrollView (Role::FrameHost)
                AXLocalFrame or AXRemoteFrame (Role::LocalFrame/RemoteFrame)

- Child frame

AccessibilityScrollView (Role::ScrollArea)
    AccessibilityRenderObject (Role::WebArea)

The normal tree-traversal methods like accessibilityChildren()
and accessibilityParentObject() no longer cross tree boundaries. That
keeps most logic within a frame, which is what we want for site isolation.

For cases where we do need to cross frame boundaries, new methods have
been added to make those cases explicit - for example when a wrapper gets
the scroll area parent, or when doing a predicate search.

No new tests, covered by hundreds of existing tests.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::crossFrameUnignoredChildren):
(WebCore::AXCoreObject::crossFrameParentObjectUnignored const):
(WebCore::AXCoreObject::crossFrameChildrenIncludingIgnored):
(WebCore::AXCoreObject::crossFrameIsAncestorOfObject const):
(WebCore::AXCoreObject::crossFrameIsDescendantOfObject const):
(WebCore::AXCoreObject::parentObjectIncludingCrossFrame const):
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isLocalFrame const):
(WebCore::Accessibility::crossFrameFindAncestor):
(WebCore::Accessibility::findAncestor):
* Source/WebCore/accessibility/AXGeometryManager.cpp:
(WebCore::AXGeometryManager::cacheRectIfNeeded):
(WebCore::AXGeometryManager::willUpdateObjectRegions):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXLoggerBase.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::~AXObjectCache):
(WebCore::AXObjectCache::focusedObjectForLocalFrame):
(WebCore::AXObjectCache::setIsolatedTreeFocusedObject):
(WebCore::AXObjectCache::getOrCreateSlow):
(WebCore::AXObjectCache::getOrCreateIsolatedTree):
(WebCore::AXObjectCache::buildIsolatedTree):
(WebCore::AXObjectCache::create):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::onEventListenerAdded):
(WebCore::AXObjectCache::onEventListenerRemoved):
(WebCore::AXObjectCache::updateLoadingProgress):
(WebCore::AXObjectCache::handleAllDeferredChildrenChanged):
(WebCore::AXObjectCache::onTextRunsChanged):
(WebCore::AXObjectCache::onPageActivityStateChange):
(WebCore::AXObjectCache::onScrollbarFrameRectChange):
(WebCore::AXObjectCache::onAccessibilityPaintFinished):
(WebCore::AXObjectCache::onFontChange):
(WebCore::AXObjectCache::onTextColorChange):
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::onSelectedTextChanged):
(WebCore::AXObjectCache::dirtyIsolatedTreeRelations):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::updateIsolatedTree const):
(WebCore::AXObjectCache::onPaint const):
(WebCore::AXObjectCache::onPaint):
(WebCore::AXObjectCache::addRelation):
(WebCore::AXObjectCache::processQueuedIsolatedNodeUpdates):
(WebCore::AXObjectCache::focusedObjectForPage): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectCacheInlines.h:
(WebCore::AXObjectCache::objectBecameIgnored):
(WebCore::AXObjectCache::objectBecameUnignored):
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::appendChildrenToArray):
(WebCore::AXSearchManager::revealHiddenMatchWithTimeout):
(WebCore::AXSearchManager::findMatchingObjectsInternal):
* Source/WebCore/accessibility/AXTreeStore.cpp:
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::roleToString):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::focusedUIElement const):
(WebCore::AccessibilityObject::focusedUIElementInAnyFrame const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityObjectInlines.h:
(WebCore::AccessibilityObject::isHidden const):
(WebCore::AccessibilityObject::isVisible const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::crossFrameParentObject const):
(WebCore::AccessibilityRenderObject::crossFrameChildObject const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityRole.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::isRoot const):
(WebCore::AccessibilityScrollView::determineAccessibilityRole):
(WebCore::AccessibilityScrollView::computeIsIgnored const):
(WebCore::AccessibilityScrollView::addLocalFrameChild):
(WebCore::AccessibilityScrollView::addChildren):
(WebCore::AccessibilityScrollView::parentObject const):
(WebCore::AccessibilityScrollView::crossFrameParentObject const):
(WebCore::AccessibilityScrollView::crossFrameChildObject const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::crossFrameSortedDescendants const):
(WebCore::Accessibility::createPlatformRoleMap):
(WebCore::AXCoreObject::sortedDescendants const): Deleted.
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper determineIsAccessibilityElement]):
(-[WebAccessibilityObjectWrapper accessibilityFocusedUIElement]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::focusedUIElementInAnyFrame const):
(WebCore::AXIsolatedObject::crossFrameParentObject const):
(WebCore::AXIsolatedObject::crossFrameChildObject const):
(WebCore::AXIsolatedObject::computeCrossFrameParentObjectWithOptionalLock const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::treeFrameCache):
(WebCore::AXIsolatedTree::createEmpty):
(WebCore::AXIsolatedTree::create):
(WebCore::AXIsolatedTree::applyPendingRootNodeLocked):
(WebCore::AXIsolatedTree::storeTree):
(WebCore::AXIsolatedTree::removeTreeForFrameID):
(WebCore::AXIsolatedTree::treeForFrameID):
(WebCore::AXIsolatedTree::treeForFrameIDAlreadyLocked):
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
(WebCore::createIsolatedObjectData):
(WebCore::AXIsolatedTree::treePageCache): Deleted.
(WebCore::AXIsolatedTree::removeTreeForPageID): Deleted.
(WebCore::AXIsolatedTree::treeForPageID): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::treeForFrameID):
(WebCore::AXIsolatedTree::treeForPageID): Deleted.
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::attachPlatformWrapper):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::addSortedObjects):
(WebCore::AXObjectCache::removeLiveRegion):
(WebCore::AXObjectCache::initializeSortedIDLists):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeNames]):
(scrollViewParent):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
(-[WebAccessibilityObjectWrapper accessibilityFocusedUIElement]):
(-[WebAccessibilityObjectWrapper accessibilityParameterizedAttributeNames]):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearAccessibilityIsolatedTree):
(WebCore::Page::doAfterUpdateRendering):
(WebCore::Page::clearAXObjectCache):
(WebCore::Page::existingAXObjectCache):
(WebCore::Page::axObjectCache):
* Source/WebCore/page/Page.h:
(WebCore::Page::existingAXObjectCache): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKAccessibilityFocusedObject):

Canonical link: <a href="https://commits.webkit.org/300491@main">https://commits.webkit.org/300491@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f18097fac241b40d3f7f843ebf34972b5ab71b07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122752 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42464 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33154 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129377 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124628 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43183 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51057 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93293 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125704 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34433 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/109898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/73932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/28052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72867 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/114906 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104133 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28265 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132104 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121281 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49697 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/37835 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/101810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50074 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106112 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/101680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25819 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/47058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/25243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46626 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49554 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55307 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151547 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/49021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38752 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50704 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
```diff
@@ -182,6 +182,16 @@ void appendPlatformProperties(AXPropertyVector& properties, OptionSet<AXProperty
+            return;
+        }
+    }
+
```
Nit: extra space here

hoffmanjoshua commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1616,6 +1621,58 @@ String AXIsolatedObject::linkRelValue() const
+
+    auto parentObjectID = *markableParentObjectID;
+
+    RefPtr parentTree = AXIsolatedTree::treeForFrameID(parentFrameID);
```
Just wanting to understand this a bit more: will this only work across local frames? If the parent tree here is in another process, will we get a tree here?

minorninth replied:
That's correct, it only works for local frames. I added a comment to the header.

hoffmanjoshua replied:
Gotcha! I don't know if this comment came through in the latest commit, looking at `AXIsolatedObject.h`

minorninth replied:
I previously added a comment to treeForFrameID in AXIsolatedTree.h and there was one in AXCoreObject.h, but I added one to AXIsolatedObject.h too, thanks.

hoffmanjoshua commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1975,6 +1992,15 @@ IsolatedObjectData createIsolatedObjectData(const Ref<AccessibilityObject>& axOb
             // FIXME: We never update this property, e.g. when the iframe is moved in the hosting web content process.
             setProperty(AXProperty::RemoteFrameOffset, object.remoteFrameOffset());
+
+            AXCoreObject* crossFrameParent = axObject->crossFrameParentObject();
```
```suggestion
            RefPtr crossFrameParent = axObject->crossFrameParentObject();
```

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3061,10 +3061,28 @@ AccessibilityObject* AccessibilityObject::elementAccessibilityHitTest(const IntP
+    return axObjectCache ? axObjectCache->focusedObjectForLocalFrame() : nullptr;
+}
+
+AccessibilityObject* AccessibilityObject::focusedUIElementInAnyFrame() const
```
Would would you think about making this more specific, like `focusedUIElementInAnyLocalFrame()`? That's with my assumption that this shouldn't work across site-isolation boundaries, but I might be misunderstanding this!

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -296,8 +367,19 @@ void AccessibilityScrollView::addChildren()
+    if (!cache)
+        return;
+
+    bool isRoot = document() == cache->document();
```
What do you think about using more specific naming here, like `isFrameRoot`? `root` is pretty overloaded already, and now that we don't have one `AXObjectCache` for a process, the root has a little different meaning than it did before.

hoffmanjoshua replied:
Would also be nice to have this as its own method, since we use it elsewhere too!

minorninth replied:
Cleaned up and replaced with isRoot throughout, and added a comment.

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -382,6 +473,58 @@ AccessibilityObject* AccessibilityScrollView::parentObject() const
+    if (frameView->frame().isMainFrame())
+        return nullptr;
+
+    // We only want to return something if this is the root of a non-root frame.
```
Do you think we could move this line up above `if (frameView->frame().isMainFrame())`?

minorninth replied:
Refactored.

hoffmanjoshua commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -760,6 +763,22 @@ class AXCoreObject : public RefCountedAndCanMakeWeakPtr<AXCoreObject> {
+    virtual AXCoreObject* crossFrameChildObject() const = 0;
+#endif
+
+    // Helpers that run on any platform and return children and parents, calling the cross frame functions if needed.
```
If these are meant to work for both Local/Remote frames, an explanation of what the different would be for the two I think could help with understanding, if that sounds reasonable?

minorninth replied:
These only walk local frames, I clarified in the comments.

The overall idea is:
(1) Keep frames as separate as possible. Normally calling accessibilityChildren() or parentObject() keeps you within the same frame. Information from one frame won't "leak" to another frame, bringing us closer to the situation when site isolation is on.
(2) At the platform wrapper layer, or to support those APIs, we need to be able to walk the "composed tree" of all local frames - so that's why we need some cross-frame helpers for element search.
(3) We can't do that for remote frames, so we'll need to rely on the platform to bridge that for remote elements.

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -355,8 +356,88 @@ AXCoreObject* AXCoreObject::firstUnignoredChild()
+
 #endif // ENABLE(INCLUDE_IGNORED_IN_CORE_AX_TREE)
 
+AXCoreObject::AccessibilityChildrenVector AXCoreObject::crossFrameUnignoredChildren(bool updateChildrenIfNeeded)
```
Until we have a callsite that actually passes a false value, maybe we can just remove this `updateChildrenIfNeeded` parameter? I've been considering trying to remove all `unignoredChildren(/* updateChildrenIfNeeded */ false)` variants because I feel like it's basically never the correct thing to do (but no need to get into that now)

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -355,8 +356,88 @@ AXCoreObject* AXCoreObject::firstUnignoredChild()
+    if (result.isEmpty()) {
+        AXCoreObject* crossFrameChild = crossFrameChildObject();
+        if (crossFrameChild)
+            result.append(*crossFrameChild);
```
```cpp
if (RefPtr crossFrameChild = crossFrameChildObject())
    result.append(crossFrameChild.releaseNonNull());
```

Satifies the safer-cpp bot while avoiding unnecessary ref-count churn

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -375,7 +456,13 @@ void AXCoreObject::verifyChildrenIndexInParent(const AccessibilityChildrenVector
+    return nextInPreOrder(updateChildrenIfNeeded, stayWithin, false);
+}
+
+AXCoreObject* AXCoreObject::nextInPreOrder(bool updateChildrenIfNeeded , AXCoreObject* stayWithin, bool crossFrame)
```
Typically, WebKit coding style recommends bool values start with `should`, `is`, `has`, `include`, etc. Maybe `includeCrossFrame`? Could also consider something like `enum class IncludeCrossFrame : bool { No, Yes }` for optimal readability at callsites without `/* includeCrossFrame */ false` or the like

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -284,12 +285,15 @@ AXObjectCache::AXObjectCache(Page& page, Document* document)
     // If loading completed before the cache was created, loading progress will have been reset to zero.
     // Consider loading progress to be 100% in this case.
-    m_loadingProgress = page.progress().estimatedProgress();
+    Page* page = localFrame.page();
+    if (page) {
+        m_loadingProgress = page->progress().estimatedProgress();
+        m_pageActivityState = page->activityState();
+    }
```
Do you have an idea of what circumstance we won't have a page? My fear here is that if we never update loading progress for some reason, VoiceOver will constantly report "loading x%", which is super annoying. And the activity state is hugely important for computing the right focused element in `WebProcess::accessibilityFocusedUIElement()`. So I could see issues happening if we don't have a page sometimes

minorninth replied:
I don't think that should ever happen unless this code is somehow called in teardown. Previously this was called with a Page but the codepath still depended on a LocalFrame existing, I just didn't want to not null-check here.

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -382,6 +473,58 @@ AccessibilityObject* AccessibilityScrollView::parentObject() const
+    if (RefPtr localFrameView = dynamicDowncast<LocalFrameView>(m_scrollView.get()))
+        ancestorElement = localFrameView->frame().ownerElement();
+
+    RefPtr ancestorDocument = ancestorElement->document();
```
Do we need a null-check on `ancestorElement`?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -1616,6 +1621,58 @@ String AXIsolatedObject::linkRelValue() const
+    return computeCrossFrameParentObjectWithOptionalLock(true);
+}
+
+AXIsolatedObject* AXIsolatedObject::computeCrossFrameParentObjectWithOptionalLock(bool acquireLock) const
```
Does anything pass a false value to this method?

minorninth replied:
Good catch, this ended up no longer being needed. I only needed treeForFrameIDAlreadyLocked

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -203,6 +204,9 @@ void AXIsolatedTree::applyPendingRootNodeLocked()
 
     if (m_pendingRootNodeID) {
+        WeakPtr cache = m_axObjectCache;
+        auto frameID = cache->frameID();
```
Is the WeakPtr local variable necessary?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1876,6 +1887,12 @@ IsolatedObjectData createIsolatedObjectData(const Ref<AccessibilityObject>& axOb
         appendBasePlatformProperties(properties, propertyFlags, axObject);
+
+        if (object.isLocalFrame()) {
+            auto frameID = downcast<AXLocalFrame>(&object)->frameID();
```
Could we use dynamicDowncast rather is + downcast?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
```diff
@@ -568,7 +572,8 @@ class AXIsolatedTree : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<AX
 
-    static HashMap<PageIdentifier, Ref<AXIsolatedTree>>& treePageCache() WTF_REQUIRES_LOCK(s_storeLock);
+    // TODO: figure out a way to enforce WTF_REQUIRES_LOCK when we might need to access it while already holding the lock
+    static HashMap<FrameIdentifier, Ref<AXIsolatedTree>>& treeFrameCache(); // WTF_REQUIRES_LOCK(s_storeLock);
```
Not sure if this would work. But maybe alongside the existing method, we add a new `static HashMap<FrameIdentifier, Ref<AXIsolatedTree>>& treeFrameCacheLocked()` method that asserts the lock is held within the function body but doesn't have `WTF_REQUIRES_LOCK(s_storeLock)` on the method definition?

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -58,8 +59,18 @@ AccessibilityScrollView::~AccessibilityScrollView()
 bool AccessibilityScrollView::isRoot() const
 {
+#if ENABLE_ACCESSIBILITY_LOCAL_FRAME
+    // Interpret this as "is this the root of the local frame"
```
Thanks for these comments 🙌🏻

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -355,8 +356,87 @@ AXCoreObject* AXCoreObject::firstUnignoredChild()
+        for (size_t i = 0; i < result.size(); i++) {
+            AXCoreObject* crossFrameChild = result[i]->crossFrameChildObject();
+            if (crossFrameChild)
+                result[i] = *crossFrameChild;
```
A bit shorter:

```cpp
if (auto* crossFrameChild = result[i]->crossFrameChildObject())
    result[i] = *crossFrameChild;
```

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -355,8 +356,87 @@ AXCoreObject* AXCoreObject::firstUnignoredChild()
+    if (!result) {
+        AXCoreObject* crossFrameParent = crossFrameParentObject();
+        if (crossFrameParent)
+            result = crossFrameParent;
```
A bit shorter, and avoids unnecessary ref-count churn since we directly return the raw pointer:

```cpp
if (auto* crossFrameParent = crossFrameParentObject())
    return crossFrameParent;
```

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -355,8 +356,87 @@ AXCoreObject* AXCoreObject::firstUnignoredChild()
+
+#if ENABLE_ACCESSIBILITY_LOCAL_FRAME
+    if (result.isEmpty()) {
+        AXCoreObject* crossFrameChild = crossFrameChildObject();
+        if (crossFrameChild)
+            result.append(*crossFrameChild);
```
A bit shorter:

```cpp
if (auto* crossFrameChild = result[i]->crossFrameChildObject())
    result.append(*crossFrameChild);
```

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -355,8 +356,87 @@ AXCoreObject* AXCoreObject::firstUnignoredChild()
+    if (!result) {
+        AXCoreObject* crossFrameParent = crossFrameParentObject();
+        if (crossFrameParent)
+            result = crossFrameParent;
```
A bit shorter, and avoids unnecessary ref-count churn since we directly return the raw pointer:

```cpp
if (auto* crossFrameParent = crossFrameParentObject())
    return crossFrameParent;
```

twilco commented on Source/WebCore/accessibility/AXLocalFrame.h:
```diff
@@ -0,0 +1,61 @@
+
+#if ENABLE_ACCESSIBILITY_LOCAL_FRAME
+    SingleThreadWeakPtr<LocalFrameView> m_localFrameView;
+    std::optional<FrameIdentifier> m_frameID { };
```
We reference `FrameIdentifier` here — do we need to import `FrameIdentifier.h`?

minorninth replied:
Ah, good call. Will save us pain when the unified build files get shuffled around in the future.

twilco commented on Source/WebCore/accessibility/AXLocalFrame.h:
```diff
@@ -0,0 +1,61 @@
+    virtual ~AXLocalFrame() = default;
+    explicit AXLocalFrame(AXID, AXObjectCache&);
+
+    AccessibilityRole determineAccessibilityRole() final { return AccessibilityRole::LocalFrame; }
```
I believe referencing a variant of `AccessibilityRole` will require including `AccessibilityRole.h`

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -284,12 +285,15 @@ AXObjectCache::AXObjectCache(Page& page, Document* document)
     // If loading completed before the cache was created, loading progress will have been reset to zero.
     // Consider loading progress to be 100% in this case.
-    m_loadingProgress = page.progress().estimatedProgress();
+    Page* page = localFrame.page();
+    if (page) {
```
A shorter option to consider:

```cpp
if (auto* page = localFrame.page()) {
```

twilco commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -523,7 +524,7 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     Document* document() const { return m_document.get(); }
     RefPtr<Document> protectedDocument() const;
-    constexpr const std::optional<PageIdentifier>& pageID() const { return m_pageID; }
+    constexpr const std::optional<FrameIdentifier>& frameID() const { return m_frameID; }
```
I think this may require including `FrameIdentifier.h` to avoid breaking non-unified builds?

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.h:
```diff
@@ -95,6 +102,9 @@ class AccessibilityScrollView final : public AccessibilityObject {
     RefPtr<AccessibilityObject> m_horizontalScrollbar;
     RefPtr<AccessibilityObject> m_verticalScrollbar;
+#if ENABLE_ACCESSIBILITY_LOCAL_FRAME
+    RefPtr<AXLocalFrame> m_localFrame;
```
I think this new AXLocalFrame object suffers the same problem that most other AccessibilityMockObject subclasses do — nothing ever calls `AXObjectCache::remove` on them, so they leak in potentially many places (but definitely the isolated tree, since that's the only mechanism by which the isolated tree knows to clean things up). Can we add a FIXME mentioning that this leaks because nothing calls `AXObjectCache::remove` on it?

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -296,8 +371,18 @@ void AccessibilityScrollView::addChildren()
+#else
     addRemoteFrameChild();
     addChild(webAreaObject());
+#endif
```
Could be nice to have a comment bracketing the compiler flag:

```cpp
#endif // ENABLE_ACCESSIBILITY_LOCAL_FRAME
```

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -382,6 +476,58 @@ AccessibilityObject* AccessibilityScrollView::parentObject() const
+    return ancestorAccessibilityObject.get();
+}
+
+AccessibilityObject* AccessibilityScrollView::crossFrameChildObject() const
+{
+    return nullptr;
+}
```
Can this be inlined in the header?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -402,6 +402,11 @@ AXIsolatedObject* AXIsolatedObject::focusedUIElement() const
+AXIsolatedObject* AXIsolatedObject::focusedUIElementInAnyLocalFrame() const
+{
+    return tree()->focusedNode().get();
+}
```
Can this be inlined in the header?

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -266,22 +271,29 @@ void AXIsolatedTree::reportLoadingProgress(double processingProgress)
-    if (RefPtr tree = treePageCache().get(pageID))
+    if (RefPtr tree = treeFrameCache().get(frameID))
+        return tree;
+    return nullptr;
```
Can this be:

```cpp
return treeFrameCache().get(frameID);
```

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -266,22 +271,29 @@ void AXIsolatedTree::reportLoadingProgress(double processingProgress)
+{
+    if (RefPtr tree = treeFrameCache().get(frameID))
         return tree;
     return nullptr;
```
Can this be:

```cpp
return treeFrameCache().get(frameID);
```

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1882,6 +1894,16 @@ IsolatedObjectData createIsolatedObjectData(const Ref<AccessibilityObject>& axOb
+            if (auto* localFrame = dynamicDowncast<AXLocalFrame>(&object)) {
+                auto frameID = localFrame->frameID();
+                if (frameID)
+                    setProperty(AXProperty::CrossFrameChildFrameID, *frameID);
```
A shorter option (I also personally like spelling out std::optional instead of auto'ing it):

```cpp
if (std::optional frameID =  localFrame->frameID())
    setProperty(AXProperty::CrossFrameChildFrameID, *frameID);
```

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -579,8 +579,12 @@ - (NSArray *)accessibilityAttributeNames
         return nil;
 
+    // TODO: we definitely don't want to do this for most ScrollViews.
+    // Figure out some more specific logic to do this only when needed.
```
> we definitely don't want to do this for most ScrollViews
What is "this"? Is it referring to the `isAttachment()` check here? I don't think attachments are always scroll views, so this comment is confusing to me. And this method is also called for all types of objects. Is there any way to clarify this comment?

minorninth replied:
Clarified, thanks. For something like a pdf, we probably do want to return accessibilityAttributeNames here. If the widget is an AccessibilityScrollView, we don't.

minorninth committed:
AX: use one AXObjectCache per frame
https://bugs.webkit.org/show_bug.cgi?id=298901
rdar://160643943

Reviewed by Tyler Wilcock and Joshua Hoffman.

Previously we had one AXObjectCache per page, spanning all frames
on that page. That design presented some challenges when trying to
implement support for remote frames - what is the root of the AXObjectCache
when the root frame is remote?

This patch refactors everything so that there's a separate AXObjectCache
for each LocalFrame. The AXObjectCache is owned by the document.

A class, AXLocalFrame, is added to parallel AXRemoteFrame.
It's a node in the accessibility tree, but its wrapper points
directly to the root of the corresponding child frame. This mimics
how AXRemoteFrame works.

Note that both with and without site isolation, the
AccessibilityScrollView class appears twice in each frame: it represents
the root of each frame, and there's also one instance of AccessibilityScrollView
for each child frame. To distinguish between these more easily, I added a new
role FrameHost for the scroll view in the parent frame.

As an example, suppose we have a simple document with an iframe in it. The
AX tree looks like this, with a total of three AccessibilityScrollView
instances.

- Parent frame

AccessibilityScrollView (Role::ScrollArea)
    AccessibilityRenderObject (Role::WebArea)
        ....
	AccessibilityRenderObject (corresponding to IFrame element, ignored)
	    AccessibilityScrollView (Role::FrameHost)
	        AXLocalFrame or AXRemoteFrame (Role::LocalFrame/RemoteFrame)

- Child frame

AccessibilityScrollView (Role::ScrollArea)
    AccessibilityRenderObject (Role::WebArea)

The normal tree-traversal methods like accessibilityChildren()
and accessibilityParentObject() no longer cross tree boundaries. That
keeps most logic within a frame, which is what we want for site isolation.

For cases where we do need to cross frame boundaries, new methods have
been added to make those cases explicit - for example when a wrapper gets
the scroll area parent, or when doing a predicate search.

No new tests, covered by hundreds of existing tests.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::crossFrameUnignoredChildren):
(WebCore::AXCoreObject::crossFrameParentObjectUnignored const):
(WebCore::AXCoreObject::crossFrameChildrenIncludingIgnored):
(WebCore::AXCoreObject::crossFrameIsAncestorOfObject const):
(WebCore::AXCoreObject::crossFrameIsDescendantOfObject const):
(WebCore::AXCoreObject::parentObjectIncludingCrossFrame const):
(WebCore::AXCoreObject::nextInPreOrder):
(WebCore::AXCoreObject::nextSiblingIncludingIgnored const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::AXCoreObject::isLocalFrame const):
(WebCore::Accessibility::crossFrameFindAncestor):
(WebCore::Accessibility::findAncestor):
* Source/WebCore/accessibility/AXGeometryManager.cpp:
(WebCore::AXGeometryManager::cacheRectIfNeeded):
(WebCore::AXGeometryManager::willUpdateObjectRegions):
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXLoggerBase.cpp:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::AXObjectCache):
(WebCore::AXObjectCache::~AXObjectCache):
(WebCore::AXObjectCache::focusedObjectForLocalFrame):
(WebCore::AXObjectCache::setIsolatedTreeFocusedObject):
(WebCore::AXObjectCache::getOrCreateSlow):
(WebCore::AXObjectCache::getOrCreateIsolatedTree):
(WebCore::AXObjectCache::buildIsolatedTree):
(WebCore::AXObjectCache::create):
(WebCore::AXObjectCache::remove):
(WebCore::AXObjectCache::onEventListenerAdded):
(WebCore::AXObjectCache::onEventListenerRemoved):
(WebCore::AXObjectCache::updateLoadingProgress):
(WebCore::AXObjectCache::handleAllDeferredChildrenChanged):
(WebCore::AXObjectCache::onTextRunsChanged):
(WebCore::AXObjectCache::onPageActivityStateChange):
(WebCore::AXObjectCache::onScrollbarFrameRectChange):
(WebCore::AXObjectCache::onAccessibilityPaintFinished):
(WebCore::AXObjectCache::onFontChange):
(WebCore::AXObjectCache::onTextColorChange):
(WebCore::AXObjectCache::onStyleChange):
(WebCore::AXObjectCache::onSelectedTextChanged):
(WebCore::AXObjectCache::dirtyIsolatedTreeRelations):
(WebCore::AXObjectCache::performDeferredCacheUpdate):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::updateIsolatedTree const):
(WebCore::AXObjectCache::onPaint const):
(WebCore::AXObjectCache::onPaint):
(WebCore::AXObjectCache::addRelation):
(WebCore::AXObjectCache::processQueuedIsolatedNodeUpdates):
(WebCore::AXObjectCache::focusedObjectForPage): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXObjectCacheInlines.h:
(WebCore::AXObjectCache::objectBecameIgnored):
(WebCore::AXObjectCache::objectBecameUnignored):
* Source/WebCore/accessibility/AXSearchManager.cpp:
(WebCore::appendChildrenToArray):
(WebCore::AXSearchManager::revealHiddenMatchWithTimeout):
(WebCore::AXSearchManager::findMatchingObjectsInternal):
* Source/WebCore/accessibility/AXTreeStore.cpp:
* Source/WebCore/accessibility/AXUtilities.cpp:
(WebCore::roleToString):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::focusedUIElement const):
(WebCore::AccessibilityObject::focusedUIElementInAnyFrame const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityObjectInlines.h:
(WebCore::AccessibilityObject::isHidden const):
(WebCore::AccessibilityObject::isVisible const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::crossFrameParentObject const):
(WebCore::AccessibilityRenderObject::crossFrameChildObject const):
* Source/WebCore/accessibility/AccessibilityRenderObject.h:
* Source/WebCore/accessibility/AccessibilityRole.h:
* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::isRoot const):
(WebCore::AccessibilityScrollView::determineAccessibilityRole):
(WebCore::AccessibilityScrollView::computeIsIgnored const):
(WebCore::AccessibilityScrollView::addLocalFrameChild):
(WebCore::AccessibilityScrollView::addChildren):
(WebCore::AccessibilityScrollView::parentObject const):
(WebCore::AccessibilityScrollView::crossFrameParentObject const):
(WebCore::AccessibilityScrollView::crossFrameChildObject const):
* Source/WebCore/accessibility/AccessibilityScrollView.h:
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::crossFrameSortedDescendants const):
(WebCore::Accessibility::createPlatformRoleMap):
(WebCore::AXCoreObject::sortedDescendants const): Deleted.
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper determineIsAccessibilityElement]):
(-[WebAccessibilityObjectWrapper accessibilityFocusedUIElement]):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::focusedUIElementInAnyFrame const):
(WebCore::AXIsolatedObject::crossFrameParentObject const):
(WebCore::AXIsolatedObject::crossFrameChildObject const):
(WebCore::AXIsolatedObject::computeCrossFrameParentObjectWithOptionalLock const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::treeFrameCache):
(WebCore::AXIsolatedTree::createEmpty):
(WebCore::AXIsolatedTree::create):
(WebCore::AXIsolatedTree::applyPendingRootNodeLocked):
(WebCore::AXIsolatedTree::storeTree):
(WebCore::AXIsolatedTree::removeTreeForFrameID):
(WebCore::AXIsolatedTree::treeForFrameID):
(WebCore::AXIsolatedTree::treeForFrameIDAlreadyLocked):
(WebCore::AXIsolatedTree::processQueuedNodeUpdates):
(WebCore::createIsolatedObjectData):
(WebCore::AXIsolatedTree::treePageCache): Deleted.
(WebCore::AXIsolatedTree::removeTreeForPageID): Deleted.
(WebCore::AXIsolatedTree::treeForPageID): Deleted.
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
(WebCore::AXIsolatedTree::treeForFrameID):
(WebCore::AXIsolatedTree::treeForPageID): Deleted.
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::attachPlatformWrapper):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXObjectCache::addSortedObjects):
(WebCore::AXObjectCache::removeLiveRegion):
(WebCore::AXObjectCache::initializeSortedIDLists):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper accessibilityAttributeNames]):
(scrollViewParent):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:]):
(-[WebAccessibilityObjectWrapper accessibilityFocusedUIElement]):
(-[WebAccessibilityObjectWrapper accessibilityParameterizedAttributeNames]):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::clearAXObjectCache):
(WebCore::Document::existingAXObjectCacheSlow const):
(WebCore::Document::axObjectCache const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearAccessibilityIsolatedTree):
(WebCore::Page::doAfterUpdateRendering):
(WebCore::Page::clearAXObjectCache):
(WebCore::Page::existingAXObjectCache):
(WebCore::Page::axObjectCache):
* Source/WebCore/page/Page.h:
(WebCore::Page::existingAXObjectCache): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKAccessibilityFocusedObject):

Canonical link: https://commits.webkit.org/300491@main
